src/core/ipns/resolver.js

```diff
@@ -116,9 +119,24 @@ class IpnsResolver {
         return callback(errcode(new Error(errMsg), 'ERR_NO_RECORD_FOUND'))
       }
 
+      // If public key was found in the routing, add it to the peer id
+      // otherwise, wait to check if it is embedded in the record.
+      if (res.pubKey) {
```

Comment:
Can this be `null` and not 404 error?

Reply:
If we fail to get the public key, but we can get the record, we should keep going, in order to try to resolve the record correctly. Two main reasons for this:

- the IPNS record will almost for sure have the public key in it, and consequently, if we end up failing to get the public key for any reason, we should try to keep going until checking if it has the public key
- https://github.com/ipfs/js-ipfs/blob/master/src/core/ipns/publisher.js#L77 in JS land, we always put the public key in the ipns record. So, we are only putting it to the routing, in order to be consistent with the old go nodes that do not embed it in the record. All the public key logic will be removed in a near future, and when `go-ipfs` does that, we will not break as we keep goind in this case. (the code should be removed in that time too).

Reply:
Once the call to get the record completes, can we check at that point if the public key is in the record, and if it is, don't bother waiting until the call to get the public key completes?

Reply:
I can change to your suggestion @dirkmc . I did not do that because doing two `dht.get` in parallel would be faster than doing both in series. But, as we will almost everytime have it in the record, I think it may be better to go with your suggestion. Let's wait for @alanshaw feedback

Reply:
I don't think it's necessary to do them in series. It would be something like:

```Javascript
let complete = 0
const onComplete = once(function () {
  // ...
})
getRecord((err, rec) => {
  complete++
  if (complete === 2 || rec.hasPubKey()) onComplete()
})
getPubkeyFromNetwork((err, pubk) => {
  complete++
  if (complete === 2) onComplete()
})
```